### PR TITLE
V6 migration guide

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [ ] Unit or Functional tests are included in the PR

**Description:**

Initial Dojo 6 migration guide.

### CODE REVIEW

File: docs/V6-Migration-Guide.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# Version 6.0.0 Migration Guide
+
+Dojo 6 has landed and contains some major ergonomic changes along with a few breaking changes that you should be aware of when migrating from version 5. As much as possible these updates should be automated by using the `@dojo/cli-upgrade-app` CLI command. When the command cannot automate the upgrade, there should be helpful hints and information in the output to guide you through the manual process.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
Dojo version 6 contains several major ergonomic changes along with a few breaking changes to be aware of when migrating from version 5. As much as possible, these updates are automated by using the `@dojo/cli-upgrade-app` CLI command. When the upgrade command cannot automate the upgrade, helpful hints and information are provided in the output to guide you through the manual upgrade process for these changes.
```

File: docs/V6-Migration-Guide.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+To install the upgrade command, run the following from the project root:
+
+```
+npm install @dojo/cli-upgrade-app --no-save
+```
+
+To perform the migration, run the following command from the project root. The migrations tool should automatically detect the necessary migration from your `package.json`.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
To perform the migration, run the following command from the project root. The Dojo migration tool should automatically detect the necessary migration from your `package.json`.
```

File: docs/V6-Migration-Guide.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+```
+dojo upgrade app
+```
+
+If you are upgrading from a version before 5.0.0, please see the [previous migration guides](./V5-Migration-Guide) for more details first.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
If you are upgrading from a version before 5.0.0, please first see the [previous migration guides](./V5-Migration-Guide) for more details.
```

File: docs/V6-Migration-Guide.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+-   The `@dojo/cli` should be updated to version 6.0.0, along with all the commands used by the project.
+-   If your project is using `@dojo/widgets` and `@dojo/interop`, these packages also require upgrading to version 6.0.0.
+
+**Note:** The migration tool may create line lengths that violate your projects linting rules, be sure to run your linter and manually fix any linting rule violations.
+
+### Breaking Changes:
+
+#### [Update to the minimum support TypeScript version](https://github.com/dojo/framework/pull/331)
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
#### [Update to the new minimum TypeScript version supported by Dojo version 6](https://github.com/dojo/framework/pull/331)
```

File: docs/V6-Migration-Guide.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+As part of the enhancements made for Dojo 6, the minimum supported has changed to `3.4.5`. This is was to support the typings required by the new functional widgets and middleware.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
As part of the improvements made for Dojo version 6, the minimum version of TypeScript has changed to `3.4.5`. This change is necessary to support the typings required by the new functional widgets and middleware.
```

File: docs/V6-Migration-Guide.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+Dojo 6 should be fully compatible with `3.4.5` onwards and updating the projects dependency should satisfy Dojo, however you may find changes are required in your project code.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
Dojo version 6 should be fully compatible with `3.4.5` and newer. Updating a project's dependency should satisfy Dojo, however you may find changes are required in your project code to meet TypeScript additional checking added in newer versions.
```

File: docs/V6-Migration-Guide.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+In addition to upgrading TypeScript, the complimentary library `tslib` should be upgraded to `~1.9.1`.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
In addition to upgrading TypeScript, the complementary library `tslib` should be upgraded to `~1.9.1`.
```

File: docs/V6-Migration-Guide.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+#### [`has`](https://github.com/dojo/framework/pull/361) and [`widget-core`](https://github.com/dojo/framework/pull/306) merged with `core`
+
+As part of Dojo 6 we have continued to re-organize `@dojo/framework`'s module structure, as a result `@dojo/framework/widget-core` and `@dojo/framework/has` have been merged with `@dojo/framework/core`.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
As part of Dojo version 6 we have continued to re-organize `@dojo/framework`'s module structure. As a result, `@dojo/framework/widget-core` and `@dojo/framework/has` have been merged with `@dojo/framework/core`.
```

File: docs/V6-Migration-Guide.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+The `@dojo/cli-upgrade-app` command should automatically migrate your projects imports, please let us know if you have any issues.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
The `@dojo/cli-upgrade-app` command should automatically migrate your project's imports. Please let us know if you encounter any issues.
```

File: docs/V6-Migration-Guide.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+#### [`@dojo/framework/widget-core/d` and `@dojo/framework/widget-core/tsx` moved to `@dojo/framework/core/vdom`](https://github.com/dojo/framework/pull/360)
+
+In addition to merging `widget-core` with `core`, the `d` and `tsx` exports have been moved into `vdom`.
+
+**Note:** There is one exception, `decorate` has been moved from `d` to `util`.
+
+The `@dojo/cli-upgrade-app` command should automatically migrate your projects imports, please let us know if you have any issues.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
The `@dojo/cli-upgrade-app` command should automatically migrate your project's imports. please let us know if you encounter any issues.
```

File: docs/V6-Migration-Guide.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+As part of Dojo 6 we have continued to re-organize `@dojo/framework`'s module structure, as a result `@dojo/framework/widget-core` and `@dojo/framework/has` have been merged with `@dojo/framework/core`.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
As part of Dojo version 6 we have continued to re-organize `@dojo/framework`'s module structure. As a result, `@dojo/framework/widget-core` and `@dojo/framework/has` have merged into `@dojo/framework/core`.
```

File: docs/V6-Migration-Guide.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+The `@dojo/cli-upgrade-app` command should automatically migrate your projects imports, please let us know if you have any issues.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
The `@dojo/cli-upgrade-app` command should automatically migrate your project's imports. Please let us know if you encounter any issues.
```

File: docs/V6-Migration-Guide.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+#### [`WNode` are no longer decorated with `bind`](https://github.com/dojo/framework/pull/290)
+
+In previous versions of Dojo virtual Dom nodes created by widgets have been decorated with additional meta data, such as `bind`. These extra properties were never meant to be exposed to the end user, however as they are added by mutating the `DNodes` it exposed these the properties inadvertently. As of Dojo 6 we no longer mutate `DNodes` to ensure that this implementation detail of the framework is exposed to the end user.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
In previous versions of Dojo, virtual DOM nodes created by widgets get decorated with additional metadata such as `bind`. These extra properties were never meant to get exposed to the end user. However, as these DOM nodes get added by mutating the `DNodes`, these properties were inadvertently exposed. As of Dojo version 6, we no longer mutate `DNodes` to ensure that this implementation detail of the framework is not exposed to the end user.
```

File: docs/V6-Migration-Guide.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+We don't envisage this change affecting many projects, however if it does please raise an issue with your use case so we can investigate an officially supported mechanism.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
We don't envisage this change affecting many projects, however, if it does please raise an issue with your use case so we can investigate an officially supported mechanism.
```

File: docs/V6-Migration-Guide.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+#### [AssertionTemplate `replace` renamed to `replaceChildren`](https://github.com/dojo/framework/pull/412)
+
+The `replace` has been renamed to `replaceChildren` with the introduction of a new API for `replace` that will replace a single node of an assertion template not target just the children.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
The `replace` method has been renamed to `replaceChildren` with the introduction of a new API for `replace` that will replace a single node of an assertion template and not target just the children.
```

File: docs/V6-Migration-Guide.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+Additionally we have deprecated passing an array of `DNode` to APIs that replace/modify children in favour of a factory function to ensure that they are immutable. The existing API is still supported but you will receive a warning and we recommend changing to use a function as the array API is likely to be removed in a future release.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
Additionally we have deprecated passing an array of `DNode` to APIs that replace/modify children in favor of a factory function to ensure that the `DNode` are immutable. The existing API is still supported, but you will receive a warning and we recommend changing to use a function as the array API is likely to be removed in a future Dojo release.
```

File: docs/V6-Migration-Guide.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+// existing API
+base.setChildren('@div', [v('span')]);
+
+// recommended API
+base.setChildren('@div', () => [v('span')]);
+```
+
+#### [Transition strategy not automatically imported](https://github.com/dojo/framework/pull/418)
+
+The transition strategy for the vdom renderer is no longer implicitly imported in `vdom`, to use `VNode` transitions that leverage `exitAnimation` or `enterAnimation` the strategy will need to be explicitly imported and passed to the applications `.mount` call.
```
Comment (dylans) [inline docs/V6-Migration-Guide.md]:
```suggestion
The transition strategy for the virtual DOM renderer is no longer implicitly imported in `vdom`. To use `VNode` transitions that leverage `exitAnimation` or `enterAnimation`, the transition strategy needs to get explicitly imported and passed to the applications `.mount` call.
```

File: docs/V6-Migration-Guide.md (hunk 13 above)
Comment (agubler) [reply docs/V6-Migration-Guide.md]: I think this "However, as these DOM nodes get" should be "However, as this metadata gets"

Comment (agubler) [reply docs/V6-Migration-Guide.md]: @dylans How about: 
```
In previous versions of Dojo, virtual DOM nodes created by widgets get decorated with additional metadata such as `bind`. These extra properties were never meant to get exposed to the end user. However, as this metadata gets added by mutating the `DNodes`, they were inadvertently exposed. As of Dojo version 6, we no longer mutate `DNodes` to ensure that this implementation detail of the framework is not exposed to the end user.
```

File: docs/V6-Migration-Guide.md [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+});
+
+// existing API
+base.setChildren('@div', [v('span')]);
+
+// recommended API
+base.setChildren('@div', () => [v('span')]);
+```
+
+#### [Router no longer uses extends `QueueingEvented`](https://github.com/dojo/framework/pull/402)
```
Comment (agubler) [inline docs/V6-Migration-Guide.md]: @dylans Added this section and the one after if you fancy having a quick review

**Pull request merged**